synapse/storage/devices.py

```diff
@@ -97,20 +97,69 @@ def get_devices_by_remote(self, destination, from_stream_id):
     def _get_devices_by_remote_txn(
         self, txn, destination, from_stream_id, now_stream_id
     ):
+        # We retrieve n+1 devices from the list of outbound pokes were n is our
+        # maximum. We then check if the very last device has the same stream_id as the
+        # second-to-last device. If so, then we ignore all devices with that stream_id
+        # and only send the devices with a lower stream_id.
+        #
+        # If when culling the list we end up with no devices afterwards, we consider the
+        # device update to be too large, and simply skip the stream_id - the rationale
+        # being that such a large device list update is likely an error.
+        #
+        # Note: The code below assumes this value is at least 1
+        maximum_devices = 100
         sql = """
-            SELECT user_id, device_id, max(stream_id) FROM device_lists_outbound_pokes
+            SELECT user_id, device_id, stream_id FROM device_lists_outbound_pokes
             WHERE destination = ? AND ? < stream_id AND stream_id <= ? AND sent = ?
-            GROUP BY user_id, device_id
-            LIMIT 20
-        """
+            ORDER BY stream_id
+            LIMIT %d
+        """ % (maximum_devices + 1)
         txn.execute(sql, (destination, from_stream_id, now_stream_id, False))
 
-        # maps (user_id, device_id) -> stream_id
-        query_map = {(r[0], r[1]): r[2] for r in txn}
-        if not query_map:
+        duplicate_updates = [r for r in txn]
+
+        # Return if there are no updates to send out
+        if len(duplicate_updates) == 0:
             return (now_stream_id, [])
 
-        if len(query_map) >= 20:
+        # Perform the equivalent of a GROUP BY
+        # Iterate through the updates list and copy any non-duplicate
+        # (user_id, device_id) entries
+        updates = [duplicate_updates[0]]
+        for i in range(1, len(duplicate_updates)):
+            update = duplicate_updates[i]
+            prev_update = duplicate_updates[i - 1]
+
+            if (update[0], update[1]) == (prev_update[0], prev_update[1]):
```

Comment:
I'm not quite following what's going on here, but I think you're assuming that duplicates can only occur on adjacent rows, which is not the case.

I suggest you just build a dict which maps from `(user_id, device_id)` to `stream_id`, and then you can iterate over the results and check the dict for each row.